### PR TITLE
Dh mathbasis

### PR DESCRIPTION
**Pull request by cossorzano**

### CODE REVIEW

File: xmipp3/protocols/protocol_angular_alignment_sph.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+import numpy as np
+from sklearn.manifold import TSNE
+
+from pyworkflow import VERSION_1_2
+
+
+class XmippProtAngularAlignmentSPH(ProtAnalysis3D):
+    """ Protocol for flexible angular alignment based on spherical harmonics. """
+    _label = 'sph angular align'
+    _lastUpdateVersion = VERSION_1_2
```
Comment (dmaluenda) [inline xmipp3/protocols/protocol_angular_alignment_sph.py]: That's not true. Use `VERSION_2_0` (also in the import)

File: xmipp3/protocols/protocol_angular_alignment_sph.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            params += ' --phaseFlipped'
+        self.runJob("xmipp_angular_sph_alignment", params, numberOfMpi=self.numberOfMpi.get())
+
+
+    def createOutputStep(self):
+        fnOut = self._getFileName('fnOut')
+        mdOut = md.MetaData(fnOut)
+        i = 0
+        for row in iterRows(mdOut):
+            coeffs = mdOut.getValue(md.MDL_SPH_COEFFICIENTS, row.getObjId())
```
Comment (dmaluenda) [inline xmipp3/protocols/protocol_angular_alignment_sph.py]: This four constants must be defined in the xmippCore repository. Please, make a pull request in the xmippCore repository where they are included.

File: xmipp3/protocols/protocol_structure_map_sph.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+    Y = U * np.sqrt(S)
+
+    return (Y[:, 0:dimensions], S)
+
+
+class XmippProtStructureMapSPH(ProtAnalysis3D):
+    """ Protocol for structure mapping based on spherical harmonics. """
+    _label = 'sph struct map'
+    _lastUpdateVersion = VERSION_1_2
```
Comment (dmaluenda) [inline xmipp3/protocols/protocol_structure_map_sph.py]: Same, use `VERSION_2_0`

File: xmipp3/protocols/protocol_volume_deform_sph.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+import pyworkflow.protocol.params as params
+from pyworkflow.em.convert import ImageHandler
+from pyworkflow.em.data import Volume
+from pyworkflow import VERSION_1_2
+
+
+class XmippProtVolumeDeformSPH(ProtAnalysis3D):
+    """ Protocol for volume deformation based on spherical harmonics. """
+    _label = 'sph volume deform'
+    _lastUpdateVersion = VERSION_1_2
```
Comment (dmaluenda) [inline xmipp3/protocols/protocol_volume_deform_sph.py]: Same

File: xmipp3/protocols/protocol_angular_alignment_sph.py (hunk 1 above)
Comment (cossorzano) [reply xmipp3/protocols/protocol_angular_alignment_sph.py]: Done

File: xmipp3/protocols/protocol_angular_alignment_sph.py (hunk 2 above)
Comment (cossorzano) [reply xmipp3/protocols/protocol_angular_alignment_sph.py]: Done

File: xmipp3/protocols/protocol_structure_map_sph.py (hunk 3 above)
Comment (cossorzano) [reply xmipp3/protocols/protocol_structure_map_sph.py]: Done

File: xmipp3/protocols/protocol_volume_deform_sph.py (hunk 4 above)
Comment (cossorzano) [reply xmipp3/protocols/protocol_volume_deform_sph.py]: Done

File: xmipp3/viewers/viewer_angular_alignment_sph.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        # ax.images.append(im)
+        # plt.show()
+
+        #AJ trying something new...
+        x = tsne2D[:,0]
+        y = tsne2D[:,1]
+
+        rangeX = np.max(x)-np.min(x)
+        rangeY = np.max(y)-np.min(y)
+        if rangeX>rangeY:
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_angular_alignment_sph.py]: `sigma = max(rangeX, rangeY) / 50`

File: xmipp3/viewers/viewer_angular_alignment_sph.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        # im.set_data(xcenters, ycenters, H)
+        # ax.images.append(im)
+        # plt.show()
+
+        #AJ trying something new...
+        x = tsne2D[:,0]
+        y = tsne2D[:,1]
+
+        rangeX = np.max(x)-np.min(x)
+        rangeY = np.max(y)-np.min(y)
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_angular_alignment_sph.py]: missing `+ 1`  ?

File: xmipp3/viewers/viewer_angular_alignment_sph.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        # define grid.
+        xi = np.linspace(min(x)-0.1, max(x)+0.1, 100)
+        yi = np.linspace(min(x)-0.1, max(x)+0.1, 100)
+
+        print(x.shape)
+        print(y.shape)
+        print(xi.shape)
+        print(yi.shape)
+
+        z = np.zeros((100, 100), float)
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_angular_alignment_sph.py]: are these 100 'the same' as those on line 111 and 112? If yes, create constant

File: xmipp3/viewers/viewer_angular_alignment_sph.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+        zSize = z.shape
+        N = len(x)
+        for c in range(zSize[1]):
+            for r in range(zSize[0]):
+                for d in range(N):
+                    z[r,c] = z[r,c] + (1.0/N) * (1.0/((2*math.pi)*sigma**2)) * math.exp(-((xi[c]-x[d])**2 + (yi[r]-y[d])**2)/(2*sigma**2))
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_angular_alignment_sph.py]: comment on the formula

File: xmipp3/viewers/viewer_angular_alignment_sph.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+        #zi = griddata((x, y), z, (xi, yi), method='cubic')
+        zMax = np.max(z)
+        z = z/zMax
+
+        # contour the gridded data, plotting dots at the randomly spaced data points.
+        CS = plt.contour(xi, yi, z, 15, linewidths=0.5, colors='k')
+        CS = plt.contourf(xi, yi, z, 15, cmap=plt.cm.jet)
+        plt.colorbar()  # draw colorbar
+        # plot data points.
+        # plt.scatter(x, y, marker='o', c='b', s=0.5)
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_angular_alignment_sph.py]: remove if not needed

File: xmipp3/viewers/viewer_angular_alignment_sph.py [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+        """
+        # Write the particles
+        prot = self.protocol
+        project = prot.getProject()
+        inputSet = prot.getInputParticles()
+        fnSqlite = prot._getTmpPath('cluster_particles.sqlite')
+        cleanPath(fnSqlite)
+        partSet = SetOfParticles(filename=fnSqlite)
+        partSet.copyInfo(inputSet)
+        for point in self.getData():
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_angular_alignment_sph.py]: `partSet = [p for p in inputSet if p.getState() == Point.SELECTED]`

File: xmipp3/viewers/viewer_angular_alignment_sph.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+        return [self.clusterWindow]
+
+    def _createCluster(self):
+        """ Create the cluster with the selected particles
+        from the cluster. This method will be called when
+        the button 'Create Cluster' is pressed.
+        """
+        # Write the particles
+        prot = self.protocol
+        project = prot.getProject()
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_angular_alignment_sph.py]: move to line 174

File: xmipp3/viewers/viewer_angular_alignment_sph.py [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+        newProt.inputNmaDimred.set(prot)
+        newProt.sqliteFile.set(fnSqlite)
+
+        project.launchProtocol(newProt)
+
+    def loadData1D(self):
+        fnOut = self.protocol._getFileName('fnOut')
+        mdOut = md.MetaData(fnOut)
+        i=0
+        for row in iterRows(mdOut):
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_angular_alignment_sph.py]: enumerate

File: xmipp3/viewers/viewer_angular_alignment_sph.py [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+                tsne1D = np.vstack((tsne1D, coeff1D))
+            i+=1
+        return tsne1D
+
+    def loadData2D(self):
+        fnOut = self.protocol._getFileName('fnOut')
+        mdOut = md.MetaData(fnOut)
+        i=0
+        for row in iterRows(mdOut):
+            coeff2D = mdOut.getValue(md.MDL_SPH_TSNE_COEFF2D, row.getObjId())
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_angular_alignment_sph.py]: There's no different in the code between loadData1D and 2D, except for this flag and tsne1/2D. 
Create common method and pass proper values.

File: xmipp3/viewers/viewer_structure_map_sph.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+    def _getVisualizeDict(self):
+        return {'doShowPlot': self._visualize}
+    
+        
+    def _visualize(self, e=None):
+        nDim = self.numberOfDimensions.get()
+        fnOutput = self.protocol._defineResultsName(nDim)
+        if not os.path.exists(fnOutput):
+            return [self.errorMessage('The necessary metadata was not produced\n'
+                                      'Execute again the protocol\n',
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_structure_map_sph.py]: execute the protocal again.

File: xmipp3/viewers/viewer_structure_map_sph.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+        count = 0
+        labels = []
+        volList, _, _ = self.protocol._iterInputVolumes()
+        for voli in volList:
+            count+=1
+            labels.append("vol_%02d"%count)
+         
+        val = 0
+        if nDim == 1:
+            AX = plt.subplot(111)
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_structure_map_sph.py]: 111?

File: xmipp3/viewers/viewer_structure_map_sph.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        # Create labels        
+        count = 0
+        labels = []
+        volList, _, _ = self.protocol._iterInputVolumes()
+        for voli in volList:
+            count+=1
+            labels.append("vol_%02d"%count)
+         
+        val = 0
+        if nDim == 1:
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_structure_map_sph.py]: create separate method for each dimension

File: xmipp3/viewers/viewer_structure_map_sph.py [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+            plt.grid(True)
+            plt.show()
+                         
+        elif nDim == 2:
+            plot = plt.scatter(coordinates[:, 0], coordinates[:, 1], marker='o', c='g')
+            plt.xlabel('Dimension 1', fontsize=11)
+            plt.ylabel('Dimension 2', fontsize=11)
+            plt.title('StructMap')
+            for label, x, y in zip(labels, coordinates[:, 0], coordinates[:, 1]):
+                plt.annotate(label,
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_structure_map_sph.py]: this annotation seems to be the same for all dimensions, please extract it (together with other shared stuff, e.g. title)

File: xmipp3/viewers/viewer_structure_map_sph.py [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+                                    xycoords='data',
+                                    xy = (x2[i], y2[i]), xytext = (x2[i]+val, y2[i]+val),
+                                    textcoords = 'data', ha = 'right',
+                                     va = 'bottom', fontsize=9,
+                                    bbox = dict(boxstyle = 'round,pad=0.3',
+                                                 fc = 'yellow', alpha = 0.3))
+                                     
+                Labels.append(label)
+                
+            def update_position(e):
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_structure_map_sph.py]: other dimenstions don't have this feature?

File: xmipp3/viewers/viewer_structure_map_sph.py [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+                    label.xytext = (x2[i],y2[i])
+                    label.update_positions(fig.canvas.get_renderer())
+                fig.canvas.draw()
+            fig.canvas.mpl_connect('button_release_event', update_position)
+            plt.show()
+        
+        return plot
+        
+    def _validate(self):
+        errors = []
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_structure_map_sph.py]: `errors = protocolviewer._validate()`

File: xmipp3/viewers/viewer_structure_map_sph.py [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+        
+        numberOfDimensions=self.numberOfDimensions.get()
+        if numberOfDimensions > 3 or numberOfDimensions < 1:
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_structure_map_sph.py]: you can use validators

File: xmipp3/viewers/viewer_volume_deform_sph.py [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+                'doShowRotation': self._doShowRotation,
+                'doShowMorphOrigRef': self._doShowMorphOrigRef,
+                'doShowMorphDeformedRef': self._doShowDeformedOrigRef
+                }
+
+
+    def _doShowStrain(self, param=None):
+
+        scriptFile = self.protocol._getPath('strain_chimera.cmd')
+        fhCmd = open(scriptFile, 'w')
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_volume_deform_sph.py]: `with open(file) as f:`

File: xmipp3/viewers/viewer_volume_deform_sph.py [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+        fnbase = removeExt(self.protocol._getFileName('fnInputVol'))
+        ext = getExt(self.protocol._getFileName('fnInputVol'))
+        fninput = abspath(fnbase + ext[0:4])
+        smprt = self.protocol.outputVolume.getSamplingRate()
+
+        fnbase2 = removeExt(self.protocol._getFileName('fnOutVol'))
+        fnStrain = abspath(fnbase2)
+
+        fhCmd.write("open %s\n" % fninput)
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_volume_deform_sph.py]: same as code above, except for the file name. Please extract function

File: xmipp3/viewers/viewer_volume_deform_sph.py [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+        fnbase = removeExt(self.protocol._getFileName('fnOutVol'))
+        ext = getExt(self.protocol._getFileName('fnOutVol'))
+        fninput = abspath(fnbase + ext[0:4])
+        smprt = self.protocol.outputVolume.getSamplingRate()
+
+        fnbase2 = removeExt(self.protocol._getFileName('fnRefVol'))
+        ext2 = getExt(self.protocol._getFileName('fnRefVol'))
+        fnref = abspath(fnbase2 + ext2[0:4])
+
+        fhCmd.write("open %s\n" % fninput)
```
Comment (DStrelak) [inline xmipp3/viewers/viewer_volume_deform_sph.py]: see my previous comment

**Pull request merged**